stackit/internal/services/cdn/customdomain/resource.go

```diff
@@ -260,7 +309,92 @@ func (r *customDomainResource) ImportState(ctx context.Context, req resource.Imp
+		managedCert = c.GetCustomDomainManagedCertificate
+
+	case cdn.PutCustomDomainResponseGetCertificateAttributeType:
+		if c == nil {
+			return Certificate{}, errors.New("input of type PutCustomDomainResponseCertificate is nil")
+		}
+		customCert = c.GetCustomDomainCustomCertificate
+		managedCert = c.GetCustomDomainManagedCertificate
+
+	default:
+		return Certificate{}, fmt.Errorf("unsupported input type: %T", certInput)
+	}
+
+	// Now we process the extracted certificates
+	if customCert != nil && customCert.Type != nil && customCert.Version != nil {
+		return Certificate{
+			Type:    *customCert.Type,
+			Version: customCert.Version, // Converts from *int64 to int
+		}, nil
+	}
+
+	if managedCert != nil && managedCert.Type != nil {
+		// The version will be the zero value for int (0), as requested
+		return Certificate{
+			Type: *managedCert.Type,
+		}, nil
+	}
+
+	return Certificate{}, errors.New("certificate structure is empty, neither custom nor managed is set")
+}
+
+// buildCertificatePayload constructs the certificate part of the payload for the API request.
+// It defaults to a managed certificate if the certificate block is omitted, otherwise it creates a custom certificate.
+func buildCertificatePayload(ctx context.Context, model *CustomDomainModel) (*cdn.PutCustomDomainPayloadCertificate, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	// If the certificate block is not specified, default to a managed certificate.
+	if model.Certificate.IsNull() {
+		managedCert := cdn.NewPutCustomDomainManagedCertificate("managed")
+		certPayload := cdn.PutCustomDomainManagedCertificateAsPutCustomDomainPayloadCertificate(managedCert)
+		return &certPayload, diags
+	}
+
+	// If the certificate block is specified, it must be a custom certificate.
+	var certModel struct {
+		Certificate types.String `tfsdk:"certificate"`
+		PrivateKey  types.String `tfsdk:"private_key"`
+		Version     types.Int64  `tfsdk:"version"`
+	}
+
+	// Unpack the Terraform object into the temporary struct.
+	respDiags := model.Certificate.As(ctx, &certModel, basetypes.ObjectAsOptions{})
+	diags.Append(respDiags...)
+	if diags.HasError() {
+		return nil, diags
+	}
+	certStr := base64.StdEncoding.EncodeToString([]byte(certModel.Certificate.ValueString()))
+	keyStr := base64.StdEncoding.EncodeToString([]byte(certModel.PrivateKey.ValueString()))
+
+	if certStr == "" || keyStr == "" {
+		diags.Append(diag.NewErrorDiagnostic("Invalid certificate or private key", "Please check if the string of the public certificate and private key in PEM format"))
+		return nil, diags
+	}
+
+	customCert := cdn.NewPutCustomDomainCustomCertificate(
+		certStr,
+		keyStr,
+		"custom",
+	)
+	certPayload := cdn.PutCustomDomainCustomCertificateAsPutCustomDomainPayloadCertificate(customCert)
+
+	return &certPayload, diags
+}
+
+func mapCustomDomainFields(customDomain *cdn.CustomDomain, model *CustomDomainModel, certificateInput interface{}) error {
```

Comment:
It would make the code cleaner, if here the whole response of `GetCustomDomain()` will be passed, like show in the suggestion. Then only the response itself and the model will be passed to this map function and probably the `normalizeCertificate()` wouldn't be needed anymore and could be removed.
With this change, the function itself needs to be adjusted as well.
```suggestion
func mapCustomDomainFields(customDomain *cdn.GetCustomDomainResponse, model *CustomDomainModel) error {
```

Reply:
Actually, I use CustomDomain instead of GetCustomDomainResponse because Create doesn't have GetCustomDomainResponse. This was already our previous solution, so I believe keeping it this way is better :)